// @teambit/documenter.ui.heading@4.1.8/dist/heading.js
import { jsx as _jsx } from "react/jsx-runtime";
import classNames from 'classnames';
import { Heading } from '@teambit/base-ui.text.heading';
import sizeStyles from './heading-sizes.module.scss';
import styles from './heading.module.scss';
/**
 * A set of concrete heading for the workspace docs.
 * H1, H2, H3, H4, H5, H6.
 *
 * To keep a consistent look to the site, headers come in a limited set of sizes.
 * While each header corresponds to a default font-size, it can be overridden using the size prop:
 */
export function H1({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h1", ...restProps, className: classNames(styles.h1, sizeStyles[size || 'lg'], className) }));
}
export function H2({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h2", ...restProps, className: classNames(styles.h2, sizeStyles[size || 'lg'], className) }));
}
export function H3({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h3", ...restProps, className: classNames(styles.h3, sizeStyles[size || 'md'], className) }));
}
export function H4({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h4", ...restProps, className: classNames(styles.h4, sizeStyles[size || 'sm'], className) }));
}
export function H5({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h5", ...restProps, className: classNames(styles.h5, sizeStyles[size || 'xs'], className) }));
}
export function H6({ size, className, ...restProps }) {
    return (_jsx(Heading, { element: "h6", ...restProps, className: classNames(styles.h6, sizeStyles[size || 'xxs'], className) }));
}
