// @teambit/component-version@1.0.3/dist/version-parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSnap = exports.isTag = exports.isHash = exports.SHORT_HASH_MINIMUM_SIZE = exports.HASH_SIZE = void 0;
const semver_1 = __importDefault(require("semver"));
const exceptions_1 = require("./exceptions");
const version_1 = require("./version");
exports.HASH_SIZE = 40;
/**
 * because the directory structure is `XX/YY....`, it needs to have at least three characters.
 */
exports.SHORT_HASH_MINIMUM_SIZE = 3;
function isLatest(versionStr) {
    return versionStr === version_1.LATEST_VERSION;
}
function isSemverValid(versionStr) {
    return Boolean(semver_1.default.valid(versionStr));
}
function returnSemver(versionStr) {
    return new version_1.Version(versionStr, false);
}
function returnLatest() {
    return new version_1.Version(null, true);
}
function returnSnap(hash) {
    return new version_1.Version(hash, false);
}
/**
 * a snap is a 40 characters hash encoded in HEX. so it can be a-f and 0-9.
 * also, for convenience, a short-hash can be used, which is a minimum of 3 characters.
 */
function isHash(str) {
    return typeof str === 'string' && isHex(str) && str.length >= exports.SHORT_HASH_MINIMUM_SIZE && str.length <= exports.HASH_SIZE;
}
exports.isHash = isHash;
/**
 * a component version can be a tag (semver) or a snap (hash)
 */
function isTag(str) {
    return typeof str === 'string' && isSemverValid(str);
}
exports.isTag = isTag;
/**
 * a component version can be a tag (semver) or a snap (hash)
 */
function isSnap(str) {
    return isHash(str) && typeof str === 'string' && str.length === exports.HASH_SIZE;
}
exports.isSnap = isSnap;
function versionParser(versionStr) {
    if (!versionStr)
        return returnLatest();
    if (isLatest(versionStr))
        return returnLatest();
    if (isSemverValid(versionStr))
        return returnSemver(versionStr);
    if (isHash(versionStr))
        return returnSnap(versionStr);
    throw new exceptions_1.InvalidVersion(versionStr.toString());
}
exports.default = versionParser;
/**
 * check if the string consists of valid hexadecimal characters
 */
function isHex(str) {
    const hexRegex = /^[0-9a-fA-F]+$/;
    return hexRegex.test(str);
}
