// @teambit/community.ui.content-tabs@1.0.3/content-tabs.tsx
import React, { useEffect } from 'react';
import type { ReactNode, HTMLAttributes } from 'react';
import { Tab } from '@teambit/ui-foundation.ui.use-box.tab';
import { useTabValue } from './use-tab-value';
// @ts-ignore
import styles from './styles.module.scss';

type TabContent = {
  title: ReactNode | string;
  body: ReactNode;
};

export type ContentTabsProps = {
  tabsContent: TabContent[];
  tabsUrlParam?: string;
} & HTMLAttributes<HTMLDivElement>;

export const ContentTabs = ({
  tabsContent,
  tabsUrlParam,
  ...rest
}: ContentTabsProps) => {
  const { activeTab, setActiveTab } = useTabValue(tabsUrlParam);

  const triggerResize = () => {
    const resizeEvent = new Event('resize');
    window.dispatchEvent(resizeEvent);

    if (document.body.dispatchEvent) {
      document.body.dispatchEvent(resizeEvent);
    }
  };

  useEffect(() => {
    const timeoutId = setTimeout(() => {
      triggerResize();
    }, 0);

    return () => clearTimeout(timeoutId);
  }, [activeTab]);

  return (
    <div {...rest}>
      <div className={styles.tabs}>
        {tabsContent.map(({ title }, index) => {
          return (
            <Tab
              key={index}
              isActive={index === activeTab}
              onClick={() => setActiveTab(index)}
            >
              {title}
            </Tab>
          );
        })}
      </div>
      {tabsContent[activeTab].body}
    </div>
  );
};
