// @teambit/component-id@1.2.2/dist/component-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentID = void 0;
const legacy_bit_id_1 = require("@teambit/legacy-bit-id");
const exceptions_1 = require("./exceptions");
class ComponentID {
    constructor(
    /**
     * legacy bit component id. avoid using it directly.
     */
    _legacy, _scope) {
        this._legacy = _legacy;
        this._scope = _scope;
        if (!_legacy.name) {
            throw new Error(`ComponentID expects to get an object with "name" prop. got ${_legacy}`);
        }
        if (!_scope && !_legacy.scope)
            throw new exceptions_1.MissingScope(_legacy);
    }
    /**
     * determine whether ID has a version.
     */
    hasVersion() {
        return this._legacy.hasVersion();
    }
    /**
     * resolves the version of the component ID.
     */
    get version() {
        return this._legacy.version;
    }
    get namespace() {
        const arr = this._legacy.name.split('/');
        return arr.slice(0, -1).join('/');
    }
    /**
     * retrieves the full name of the component including its namespace.
     */
    get fullName() {
        return this._legacy.name;
    }
    /**
     * resolves the name of the component.
     */
    get name() {
        const arr = this._legacy.name.split('/');
        return arr[arr.length - 1];
    }
    /**
     * return the scope if included in the ID.
     */
    get scope() {
        const scope = this._legacy.scope;
        if (scope)
            return scope;
        if (!this._scope)
            throw new Error('scope cannot be undefined');
        return this._scope;
    }
    /**
     * get a new component ID instance with given scope.
     * in case "undefined"/"null" is passed, the current scope becomes defaultScope
     */
    changeScope(scopeName) {
        const legacyId = this._legacy.changeScope(scopeName);
        return ComponentID.fromLegacy(legacyId, this.scope);
    }
    /**
     * relevant for new components where the legacyBitId.scope is empty.
     * returns a new ComponentId whose bitId instance is the same, but the `_scope` is the given scopeName.
     */
    changeDefaultScope(scopeName) {
        return ComponentID.fromLegacy(this._legacy, scopeName);
    }
    changeVersion(version) {
        const legacyId = this._legacy.changeVersion(version);
        return ComponentID.fromLegacy(legacyId, this.scope);
    }
    isEqual(id, opts) {
        return ComponentID.isEqual(this, id, opts);
    }
    isEqualWithoutVersion(id) {
        return this.isEqual(id, { ignoreVersion: true });
    }
    isLocal(scopeName) {
        return this._legacy.isLocal(scopeName);
    }
    /**
     * do not trust this data to determine whether a component is exported to a remote or not.
     * use workspace.isExported() or scope.isExported() instead.
     */
    hasScope() {
        return this._legacy.hasScope();
    }
    serialize() {
        return this.toObject();
    }
    static deserialize(id) {
        return ComponentID.fromObject(id);
    }
    static isValidVersion(version) {
        return legacy_bit_id_1.BitId.isValidVersion(version);
    }
    static getVersionFromString(id) {
        return id.split(legacy_bit_id_1.VERSION_DELIMITER)[1];
    }
    static getStringWithoutVersion(id) {
        return id.split(legacy_bit_id_1.VERSION_DELIMITER)[0];
    }
    /**
     * examples:
     * 1.0.0 => null
     * 1.0.0-dev.1 => ['dev', 1]
     * 1.0.0-dev.1.alpha.2 => ['dev', 1, 'alpha', 2]
     * 1.0.0-0 => [0]
     */
    getVersionPreReleaseData() {
        return this._legacy.getVersionPreReleaseData();
    }
    /**
     * serialize a component ID without its version.
     */
    toStringWithoutVersion() {
        let id = this._legacy;
        if (this._scope && !this._legacy.scope) {
            id = id.changeScope(this._scope);
        }
        return id.toStringWithoutVersion();
    }
    hasSameVersion(id) {
        return this._legacy.hasSameVersion(id._legacy);
    }
    /**
     * serialize the component ID.
     */
    toString(opts = {}) {
        let id = this._legacy;
        if (this._scope && !this._legacy.scope) {
            id = id.changeScope(this._scope);
        }
        const idStr = id.toString(false, opts.ignoreVersion);
        if (opts.fsCompatible)
            return idStr.replace(/\//g, '_').replace(/\./g, '_').replace(/-/g, '_');
        return idStr;
    }
    toObject() {
        const object = this._legacy.serialize();
        if (!object.scope) {
            object.scope = this.scope;
        }
        // TODO - TS does not realize object.scope now has a value
        return object;
    }
    clone() {
        return ComponentID.fromLegacy(this._legacy.clone(), this._scope);
    }
    /**
     * generate a component ID from a string. Returns undefined if input is malformed
     */
    static tryFromString(idStr, scope) {
        try {
            return ComponentID.fromString(idStr, scope);
        }
        catch (_a) {
            return undefined;
        }
    }
    static isValidCompIdStr(idStr) {
        return Boolean(ComponentID.tryFromString(idStr));
    }
    /**
     * generate a component ID from a string.
     */
    static fromString(idStr, scope) {
        const legacyId = legacy_bit_id_1.BitId.parse(idStr, true);
        if (!scope && !legacyId.scope)
            throw new exceptions_1.MissingScope(idStr);
        return new ComponentID(legacyId, scope);
    }
    /**
     * @deprecated
     * please make sure not to use this function. it is deprecated and its usage is forbidden
     * and could potentially cause many different bugs across the system.
     */
    static fromLegacyString(idStr, scope) {
        const legacyId = legacy_bit_id_1.BitId.parse(idStr, false);
        return new ComponentID(legacyId, scope);
    }
    /** deserialize a component id from raw object */
    static fromObject(object, scope) {
        if (object instanceof ComponentID)
            throw new Error(`ComponentID.fromObject expect to get an object, got an instance of ComponentID: ${object.toString()}`);
        return ComponentID.fromLegacy(new legacy_bit_id_1.BitId(object), scope);
    }
    /**
     * check if object can be correctly deserialized to be a ComponentID
     */
    static isValidObject(o) {
        return typeof o === 'object' && typeof o.name === 'string' && typeof o.scope === 'string';
        // consider validating values with regex
    }
    static isEqual(a, b, opts = {}) {
        if (!a && !b)
            return true;
        if (!a || !b)
            return false;
        const result = a.scope === b.scope &&
            a.toString({ ignoreVersion: opts.ignoreVersion }) === b.toString({ ignoreVersion: opts.ignoreVersion });
        if (opts.ignoreVersion) {
            return result;
        }
        return result && a.hasSameVersion(b);
    }
    static isEqualObj(a, b, opts = {}) {
        if (!a && !b)
            return true;
        if (!a || !b)
            return false;
        let result = a.scope === b.scope && a.name === b.name;
        if (!opts.ignoreVersion) {
            result = result && a.version === b.version;
        }
        return result;
    }
    /**
     * create a `ComponentID` instance from the legacy `BitId`.
     */
    static fromLegacy(legacyId, scope) {
        if (!scope && !legacyId.scope)
            throw new exceptions_1.MissingScope(legacyId);
        return new ComponentID(legacyId, scope);
    }
    static sortIds(ids) {
        return ids.sort((a, b) => a.toString().localeCompare(b.toString()));
    }
}
exports.ComponentID = ComponentID;
