// @teambit/documenter.markdown.heading@0.1.10/dist/create-heading.composition.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createHeading, h1 as H1, h5 as H5 } from './create-heading';
export const CreateH1Example = () => {
    const Heading = createHeading('lg');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H1 size" });
};
export const CreateH2Example = () => {
    const Heading = createHeading('md');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H2 size" });
};
export const CreateH3Example = () => {
    const Heading = createHeading('sm');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H3 size" });
};
export const CreateH4Example = () => {
    const Heading = createHeading('xs');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H4 size" });
};
export const CreateH5Example = () => {
    const Heading = createHeading('xxs');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H5 size" });
};
export const CreateH6Example = () => {
    const Heading = createHeading('xxs');
    return _jsx(Heading, { "data-testid": "test-create-heading", children: "H6 size" });
};
export const LinkedH1Example = () => {
    return _jsx(H1, { link: "link", children: "Linked H1" });
};
export const LinkedH5Example = () => {
    return _jsx(H5, { link: "link", children: "Linked H5" });
};
