// @teambit/react.modules.dom-to-react@0.2.0/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fiberToPrototype = exports.toRootFiber = exports.domToFiber = exports.toRootElement = exports.domToReacts = exports.domToReact = void 0;
var dom_to_react_1 = require("./dom-to-react");
Object.defineProperty(exports, "domToReact", { enumerable: true, get: function () { return dom_to_react_1.domToReact; } });
Object.defineProperty(exports, "domToReacts", { enumerable: true, get: function () { return dom_to_react_1.domToReacts; } });
Object.defineProperty(exports, "toRootElement", { enumerable: true, get: function () { return dom_to_react_1.toRootElement; } });
var dom_to_fiber_1 = require("./dom-to-fiber");
Object.defineProperty(exports, "domToFiber", { enumerable: true, get: function () { return dom_to_fiber_1.domToFiber; } });
Object.defineProperty(exports, "toRootFiber", { enumerable: true, get: function () { return dom_to_fiber_1.toRootFiber; } });
Object.defineProperty(exports, "fiberToPrototype", { enumerable: true, get: function () { return dom_to_fiber_1.fiberToPrototype; } });
