// @teambit/react.ui.component-highlighter@0.2.4/children-highlighter/children-highlighter.composition.tsx
import React from 'react';
// import { EnterpriseOffering } from '@teambit/evangelist.pages.enterprise-offering';
import { MockedComponentWithMeta } from '@teambit/react.ui.highlighter.component-metadata.bit-component-meta';
import { ExcludeHighlighter } from '../ignore-highlighter';
import { ChildrenHighlighter } from './children-highlighter';
import { MockTarget } from '../mock-component';

export const ChildrenHighlighterPreview = () => {
  return (
    // highlighter runs in compositions, therefor should not have our font
    <ChildrenHighlighter style={{ padding: 40, minWidth: 200, fontFamily: 'sans-serif' }}>
      <MockedComponentWithMeta>target #1</MockedComponentWithMeta>
      <br />
      <br />
      <br />
      <br />
      <MockedComponentWithMeta>target #2</MockedComponentWithMeta>
    </ChildrenHighlighter>
  );
};

export const ChildrenHighlighterWithCustomColors = () => {
  return (
    <ChildrenHighlighter
      style={{ padding: 40, minWidth: 200, color: 'yellow', fontFamily: 'sans-serif' }}
      bgColor="cornflowerblue"
      bgColorHover="blue"
      bgColorActive="DarkSlateBlue"
    >
      <MockedComponentWithMeta>hover here</MockedComponentWithMeta>
      <br />
      <br />
      <br />
      <MockedComponentWithMeta>also here</MockedComponentWithMeta>
    </ChildrenHighlighter>
  );
};

export const ChildrenHighlighterInsideIgnore = () => {
  return (
    <ExcludeHighlighter>
      <ChildrenHighlighter style={{ fontFamily: 'sans-serif' }}>
        Multi Highlighter should still work when inside <code>{'<ExcludeHighlighter>'}</code>
        <br />
        It should only skip exclusion zones inside of it.
        <br />
        <br />
        <br />
        <MockedComponentWithMeta>hover here</MockedComponentWithMeta>
        <br />
        <br />
        <br />
        <MockedComponentWithMeta>also here</MockedComponentWithMeta>
      </ChildrenHighlighter>
    </ExcludeHighlighter>
  );
};

export const ChildrenHighlighterWithRule = () => {
  const rule = '#someSubTree *';

  return (
    <ChildrenHighlighter rule={rule} style={{ minWidth: 300, fontFamily: 'sans-serif' }}>
      <div>
        element filter: <code>"{rule}"</code>
      </div>
      <br />
      <MockedComponentWithMeta>no highlighter</MockedComponentWithMeta>
      <br />
      <br />
      <br />
      <div id="someSubTree">
        <MockedComponentWithMeta>this will be highlighted</MockedComponentWithMeta>
      </div>
    </ChildrenHighlighter>
  );
};

export const ChildrenHighlighterWithComponentRule = () => {
  const filterTarget = 'teambit.design/ui/mock-target';

  return (
    <ChildrenHighlighter componentRule={filterTarget} style={{ minWidth: 300, fontFamily: 'sans-serif' }}>
      <div>
        component filter: <code>"{filterTarget}"</code>
      </div>
      <br />
      <MockedComponentWithMeta>no highlighter</MockedComponentWithMeta>
      <br />
      <br />
      <br />
      <MockTarget style={{ display: 'inline-block', margin: '0 8px' }}>this will be highlighted</MockTarget>
    </ChildrenHighlighter>
  );
};

// export const HighlightingAllElementsInTheEnterprisePage = () => {
//   return (
//     <ChildrenHighlighter>
//       <EnterpriseOffering style={{ height: 300 }} />
//     </ChildrenHighlighter>
//   );
// };
