// @bitdev/react.generators.react-starters@1.1.1/starters/react-workspace/react-workspace-starter.ts
import { WorkspaceContext, WorkspaceTemplate } from '@teambit/generator';
import { ComponentID } from '@teambit/component-id';
import { generateFiles as generateCommonFiles } from '../../workspace-common';
import { DEFAULT_SCOPE } from '../../workspace-common/constants';

export type ReactComponentTemplateOptions = {
  /**
   * name of the template
   */
  name?: string;

  /**
   * description of the template.
   */
  description?: string;

  /**
   * hide the template from the `bit templates` command.
   */
  hidden?: boolean;
};

export class ReactWorkspaceStarter implements WorkspaceTemplate {
  constructor(
    readonly name = 'react',
    readonly description = 'React workspace with demo components',
    readonly hidden = false
  ) {}

  async generateFiles(context: WorkspaceContext) {
    const appId = this.getAppId(context);
    
    return generateCommonFiles(context, {
      [appId.toStringWithoutVersion()]: {} 
    }, ['bitdev.node/node-env']);
  }

  getAppId(context: WorkspaceContext) {
    const projectName = context.name;
    const defaultScope = context.defaultScope || DEFAULT_SCOPE;
    // const [owner] = defaultScope.split('.');

    return ComponentID.fromString(`${defaultScope}/${projectName}`);
  }

  fork(context: WorkspaceContext) {
    const projectName = context.name;
    const appId = this.getAppId(context);

    return [
      {
        id: 'acme.corporate/corporate-website',
        targetName: `${projectName}-web`
      },
      {
        id: 'acme.corporate/corporate-service',
        targetName: `${projectName}-service`,
      },
      {
        id: 'acme.corporate/acme-corporate',
        targetName: appId.fullName
      },
      {
        id: 'acme.corporate/ui/announcements',
        targetName: `ui/announcements`
      },
      {
        id: 'acme.corporate/hooks/use-announcements',
        targetName: 'hooks/use-announcements',
      },
      {
        id: 'acme.corporate/entities/announcement',
        targetName: 'entities/announcement'
      },
    ];
  }

  static from(options: ReactComponentTemplateOptions = {}) {
    return () =>
      new ReactWorkspaceStarter(
        options.name,
        options.description,
        options.hidden
      );
  }
}
