// @teambit/component-id@1.2.0/dist/component-id-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentIdList = void 0;
const lodash_1 = require("lodash");
const component_id_1 = require("@teambit/component-id");
const component_version_1 = require("@teambit/component-version");
class ComponentIdList extends Array {
    toStringArray() {
        return this.map((componentId) => componentId.toString());
    }
    has(componentId) {
        return Boolean(this.search(componentId));
    }
    hasWithoutVersion(componentId) {
        return Boolean(this.searchWithoutVersion(componentId));
    }
    hasWithoutScopeAndVersion(componentId) {
        return Boolean(this.searchWithoutScopeAndVersion(componentId));
    }
    search(componentId) {
        return this.find((id) => id.fullName === componentId.fullName &&
            id.scope === componentId.scope &&
            id._legacy.hasSameVersion(componentId._legacy));
    }
    searchWithoutVersion(componentId) {
        return this.find((id) => id.fullName === componentId.fullName && id.scope === componentId.scope);
    }
    searchWithoutScopeAndVersion(componentId) {
        return this.find((id) => id.fullName === componentId.fullName);
    }
    searchStrWithoutVersion(idStr) {
        return this.find((id) => id.toStringWithoutVersion() === idStr);
    }
    filterExact(componentId) {
        return this.filter((id) => id.fullName === componentId.fullName && id.scope === componentId.scope && id.hasSameVersion(componentId));
    }
    filterWithoutVersion(componentId) {
        return this.filter((id) => id.fullName === componentId.fullName && id.scope === componentId.scope);
    }
    removeIfExist(componentId) {
        return ComponentIdList.fromArray(this.filter((id) => !id.isEqual(componentId)));
    }
    toObject() {
        return this.reduce((acc, componentId) => {
            acc[componentId.toString()] = componentId;
            return acc;
        }, {});
    }
    toString() {
        return this.map((id) => id.toString()).join(', ');
    }
    toGroupByIdWithoutVersion() {
        return this.reduce((acc, current) => {
            const idStrWithoutVer = current.toStringWithoutVersion();
            if (acc[idStrWithoutVer])
                acc[idStrWithoutVer].push(current);
            else
                acc[idStrWithoutVer] = new ComponentIdList(current);
            return acc;
        }, {});
    }
    toGroupByScopeName() {
        return this.reduce((acc, current) => {
            const scopeName = current.scope;
            if (acc[scopeName])
                acc[scopeName].push(current);
            else
                acc[scopeName] = new ComponentIdList(current);
            return acc;
        }, {});
    }
    findDuplicationsIgnoreVersion() {
        const duplications = {};
        this.forEach((id) => {
            const sameIds = this.filterWithoutVersion(id);
            if (sameIds.length > 1) {
                duplications[id.toStringWithoutVersion()] = sameIds;
            }
        });
        return duplications;
    }
    add(componentIds) {
        componentIds.forEach((componentId) => {
            if (!this.search(componentId))
                this.push(componentId);
        });
    }
    static fromArray(ids) {
        // don't do `new componentIds(...ids);`, it'll throw "Maximum call stack size exceeded" for large number if ids.
        const componentIds = new ComponentIdList();
        ids.forEach((id) => componentIds.push(id));
        return componentIds;
    }
    static uniqFromArray(componentIds) {
        const uniq = (0, lodash_1.uniqBy)(componentIds, (id) => id.toString());
        return ComponentIdList.fromArray(uniq);
    }
    static fromStringArray(idsStr = []) {
        return ComponentIdList.fromArray(idsStr.map((id) => component_id_1.ComponentID.fromString(id)));
    }
    throwForDuplicationIgnoreVersion() {
        this.forEach((componentId) => {
            const found = this.filterWithoutVersion(componentId);
            if (found.length > 1) {
                throw new Error(`componentIds has "${componentId.toStringWithoutVersion()}" duplicated as following:
${found.map((id) => id.toString()).join('\n')}`);
            }
        });
    }
    toVersionLatest() {
        return ComponentIdList.uniqFromArray(this.map((id) => id.changeVersion(component_version_1.LATEST_VERSION)));
    }
    clone() {
        return ComponentIdList.fromArray(this.map((id) => id.clone()));
    }
}
exports.ComponentIdList = ComponentIdList;
