// @teambit/documenter.ui.paragraph@4.1.8/dist/paragraph.composition.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Paragraph } from './paragraph';
// tODO - make text stay single line
export const ParagraphXxl = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "xxl", children: "xxl text" }) }));
};
export const ParagraphXl = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "xl", children: "xl text" }) }));
};
export const ParagraphLg = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "lg", children: "lg text" }) }));
};
export const ParagraphMd = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "md", children: "md text" }) }));
};
export const ParagraphSm = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "sm", children: "sm text" }) }));
};
export const ParagraphXs = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "xs", children: "xs text" }) }));
};
export const ParagraphXxs = () => {
    return (_jsx("div", { children: _jsx(Paragraph, { size: "xxs", children: "xxs text" }) }));
};
