// @bitdev/react.generators.react-templates@4.0.2/react-env-template/react-env-template.ts
import {
  ComponentContext,
  ComponentFile,
  ComponentTemplate
} from '@teambit/generator';
import { indexFile } from './files/index-file';
import { ReactEnv } from './files/react-env-bit-env';
import { envJson } from './files/env-jsonc';
import { eslintConfigFile } from './files/config/eslintrc';
import { tsConfigFile } from './files/config/tsconfig.json';
// import { jestConfigFile } from './files/config/jest.config';
import { prettierConfigFile } from './files/config/prettier.config';
// import { webpackTransformerFile } from './files/config/webpack-transformer';
import { mounterFile } from './files/preview/mounter';
import { docsFile } from './files/preview/docs';
// import { hostDependenciesFile } from './files/preview/host-dependencies';
import { envDtsFile } from './files/types/env';
import { envDocsFile } from './files/docs';
import { viteConfigFile } from './files/config/vite.config';
import { vitestConfigFile } from './files/config/vitest.config';
import { vitestSetupFile } from './files/config/vitest.setup';


export type ReactEnvTemplateOptions = {
  /**
   * name of the template
   */
  name?: string;

  /**
   * description of the template.
   */
  description?: string;

  /**
   * hide the template from the `bit templates` command.
   */
  hidden?: boolean;
};

export class ReactEnvTemplate implements ComponentTemplate {
  constructor(
    readonly name = 'react-env',
    readonly description = 'create a custom React development environment',
    readonly hidden = false,
    readonly isEnv = true,
    readonly installMissingDependencies = true,
    readonly env = 'bitdev.general/envs/bit-env'
  ) {}

  generateFiles(context: ComponentContext): ComponentFile[] {
    return [
      indexFile(context),
      mounterFile(),
      docsFile(),
      // hostDependenciesFile(),
      ReactEnv(context),
      envJson(),
      eslintConfigFile(),
      envDocsFile(context),
      tsConfigFile(),
      // jestConfigFile(),
      viteConfigFile(),
      vitestConfigFile(),
      prettierConfigFile(),
      vitestSetupFile(),
      // webpackTransformerFile(),
      envDtsFile(),
    ];
  }

  static from(options: ReactEnvTemplateOptions = {}) {
    return () =>
      new ReactEnvTemplate(options.name, options.description, options.hidden);
  }
}
