// @bitdev/react.generators.react-templates@4.0.2/dist/react-env-template/react-env-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactEnvTemplate = void 0;
const index_file_1 = require("./files/index-file");
const react_env_bit_env_1 = require("./files/react-env-bit-env");
const env_jsonc_1 = require("./files/env-jsonc");
const eslintrc_1 = require("./files/config/eslintrc");
const tsconfig_json_1 = require("./files/config/tsconfig.json");
// import { jestConfigFile } from './files/config/jest.config';
const prettier_config_1 = require("./files/config/prettier.config");
// import { webpackTransformerFile } from './files/config/webpack-transformer';
const mounter_1 = require("./files/preview/mounter");
const docs_1 = require("./files/preview/docs");
// import { hostDependenciesFile } from './files/preview/host-dependencies';
const env_1 = require("./files/types/env");
const docs_2 = require("./files/docs");
const vite_config_1 = require("./files/config/vite.config");
const vitest_config_1 = require("./files/config/vitest.config");
const vitest_setup_1 = require("./files/config/vitest.setup");
class ReactEnvTemplate {
    constructor(name = 'react-env', description = 'create a custom React development environment', hidden = false, isEnv = true, installMissingDependencies = true, env = 'bitdev.general/envs/bit-env') {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.isEnv = isEnv;
        this.installMissingDependencies = installMissingDependencies;
        this.env = env;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, mounter_1.mounterFile)(),
            (0, docs_1.docsFile)(),
            // hostDependenciesFile(),
            (0, react_env_bit_env_1.ReactEnv)(context),
            (0, env_jsonc_1.envJson)(),
            (0, eslintrc_1.eslintConfigFile)(),
            (0, docs_2.envDocsFile)(context),
            (0, tsconfig_json_1.tsConfigFile)(),
            // jestConfigFile(),
            (0, vite_config_1.viteConfigFile)(),
            (0, vitest_config_1.vitestConfigFile)(),
            (0, prettier_config_1.prettierConfigFile)(),
            (0, vitest_setup_1.vitestSetupFile)(),
            // webpackTransformerFile(),
            (0, env_1.envDtsFile)(),
        ];
    }
    static from(options = {}) {
        return () => new ReactEnvTemplate(options.name, options.description, options.hidden);
    }
}
exports.ReactEnvTemplate = ReactEnvTemplate;
