// @bitdev/react.generators.react-starters@1.1.1/dist/starters/react-workspace/react-workspace-starter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactWorkspaceStarter = void 0;
const component_id_1 = require("@teambit/component-id");
const workspace_common_1 = require("../../workspace-common");
const constants_1 = require("../../workspace-common/constants");
class ReactWorkspaceStarter {
    constructor(name = 'react', description = 'React workspace with demo components', hidden = false) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
    }
    async generateFiles(context) {
        const appId = this.getAppId(context);
        return (0, workspace_common_1.generateFiles)(context, {
            [appId.toStringWithoutVersion()]: {}
        }, ['bitdev.node/node-env']);
    }
    getAppId(context) {
        const projectName = context.name;
        const defaultScope = context.defaultScope || constants_1.DEFAULT_SCOPE;
        // const [owner] = defaultScope.split('.');
        return component_id_1.ComponentID.fromString(`${defaultScope}/${projectName}`);
    }
    fork(context) {
        const projectName = context.name;
        const appId = this.getAppId(context);
        return [
            {
                id: 'acme.corporate/corporate-website',
                targetName: `${projectName}-web`
            },
            {
                id: 'acme.corporate/corporate-service',
                targetName: `${projectName}-service`,
            },
            {
                id: 'acme.corporate/acme-corporate',
                targetName: appId.fullName
            },
            {
                id: 'acme.corporate/ui/announcements',
                targetName: `ui/announcements`
            },
            {
                id: 'acme.corporate/hooks/use-announcements',
                targetName: 'hooks/use-announcements',
            },
            {
                id: 'acme.corporate/entities/announcement',
                targetName: 'entities/announcement'
            },
        ];
    }
    static from(options = {}) {
        return () => new ReactWorkspaceStarter(options.name, options.description, options.hidden);
    }
}
exports.ReactWorkspaceStarter = ReactWorkspaceStarter;
